framework/src/iotbus/iotbus_gpio.c

```diff
@@ -375,93 +394,164 @@ int iotbus_gpio_unregister_cb(iotbus_gpio_context_h dev)
 	return IOTBUS_ERROR_NONE;
 }
 
-#ifdef CONFIG_IOTDEV
-int iotbus_gpio_set_interrupt(iotbus_gpio_context_h dev, iotbus_int_type_e int_type, iotbus_gpio_cb cb)
+static pthread_addr_t gpio_intr_thread(pthread_addr_t arg)
 {
-	int fd, ret, edge;
+	struct intr_args *param = (struct intr_args *)arg;
+	iotbus_gpio_context_h dev = (iotbus_gpio_context_h)param->dev;
 	struct _iotbus_gpio_s *handle;
-	struct iotbus_int_info_s info = { 0, };
+	
+	siginfo_t info;
+	sigset_t sig_set;
+	uint32_t val;
 
-	handle = (struct _iotbus_gpio_s *)dev->handle;
-	handle->cb = cb;
+	sigemptyset(&sig_set);
+	sigaddset(&sig_set, SIG_IOTBUS_GPIO_RISING);
+	sigaddset(&sig_set, SIG_IOTBUS_GPIO_FALLING);
 
-	info.handle = dev;
-	info.pin_type = IOTBUS_GPIO;
-	info.int_type = int_type;
-	info.pid = iotapi_get_pid();
-	
-	fd = open(IOTBUS_SIGPATH, O_NONBLOCK);
-	if (fd < 0) {
-		ret = get_errno();
-		ibdbg("Open iotsig fail[%d, %d]\n", fd, ret);
-		return IOTBUS_ERROR_DEVICE_FAIL;
-	}
+	pthread_sigmask(SIG_BLOCK, &sig_set, NULL);
 
-	if (int_type == IOTBUS_GPIO_FALLING) {
-		edge = GPIO_EDGE_FALLING;
-	} else if (int_type == IOTBUS_GPIO_EDGE_RISING) {
-		edge = GPIO_EDGE_RISING;
-	} else {
-		edge = GPIO_EDGE_NONE;
+	handle = (struct _iotbus_gpio_s *)dev->handle;
+	if (handle->cb == NULL) {
+		return NULL;
 	}
 
-	ret = ioctl(fd, IOTBUS_INTR_REGISTER, (unsigned long)&info);
-	if (ret < 0) {
-		ibdbg("Fail to register %d\n", int_type);
-		close(fd);
-		return IOTBUS_ERROR_DEVICE_FAIL;
-	}
-	ret = ioctl(handle->fd, GPIOIOC_SET_INTERRUPT, edge);
-	if (ret < 0) {
-		ibdbg("Fail to set %d\n", int_type);
-		close(fd);
-		return IOTBUS_ERROR_DEVICE_FAIL;
+	while (1) {
+		sigwaitinfo(&sig_set, &info);
+
+#ifdef CONFIG_IOTBUS_INTERRUPT_TIMESTAMP
+#ifdef CONFIG_CAN_PASS_STRUCTS
```

Comment:
I think we don't need val variable. How about below?

`#ifdef CONFIG_CAN_PASS_STRUCTS`
`handle->timestamp = info.si_value.sival_int;`
`#else`
`handle->timestamp = info.si_value;`
`#endif`